CIPs/0010.md

```diff
@@ -0,0 +1,84 @@
+- Acts as the canonical account for the Proof-of-Stake system (aka LockedGold account)
+- Keeps track of authorized signer keys in the PoS system (validator, voter, and attestation signers)
+- Stores basic profile information such as a name and mapping to `dataEncryptionKey` and `walletAddress`
+- Entry into an account's off-chain metadata
+
+Over the past months, it has become noticeable that especially the authorization of signer keys by the "account key" is a very useful method to authenticate the "account owner" without having to unnecessarily bring the "account key" online which is the most privileged, and thus most important to secure. Example of such cases have been:
+
+- The signing of the off-chain data should ideally not happen with the account key. During the development of the `ReleaseGold` smart contract it has been found that smart contracts as accounts cannot sign metadata. As a workaround, verifying metadata has been adjusted to allow any authorized signer to sign the payload, which violates the intention behind having well scoped authorized signers in the first place.
+- As part of ODIS, accounts have to authenticate with the ODIS signers for rate-limiting purposes. To do that, the "account key" or the "authorized DEK" has to sign the request which also violates the intention of keeping keys that can move funds as cold as possible.
+- For other applications such as CIP8 or other down-the-road applications, specific keys are going to be useful. Additionally, the `dataEncryptionKey` seems out of place without useful segmentation ability, despite it being effectively also an authorized key (if not signer).
+
+Allowing the segmentation of domains for a Celo account (i.e. an account on `Accounts.sol`) gives Celo stakeholders more flexibility in their custody setup. It allows sensitive domains (like moving balances) to be only accessible by keys that are "cold", i.e. harder to compromise, while making less sensitive domains more user friendly. Expanding such segmentation aids the user experience, both of interacting with the Celo Core Contracts themselves (via authorized smart contracts), as well as dapps that currently would need to copy this authorization scheme to receive similar benefits for their users. While this CIP does not specify further applications, per-dapp-authorizations can become practical.
+
+## Specification
+
+The proposed changes to `Accounts.sol`:
+
+- Add a `signerAuthorizations` struct:
+    ```
+    struct SignerAuthorization {
+      address signer
+      bool authorized
+    }
+    mapping(address => mapping(bytes32 => SignerAuthoriation)) public signerAuthorizations
+    ```
+- Add generalized `authorizeSigner(address signer, string calldata role)` and `completeSignerAuthorization(address account, string calldata role)` for smart contract authorizations (`authorizeSigner` creates a pending authorization that `completeSignerAuthorization` completes)
```

Comment:
I take it that the `authorized` field would be used for this, with it set to `false` until the signer calls `completeSignerAuthorization()`.  If that's so, perhaps it would be better called `completed` rather than `authorized`, aligning the field name with the name of the method that sets it to true.  Granted, there is also `authorizeSignerWithSignature()`, which doesn't mention completion, but still I would have a slight preference for `completed`.

And a question: say I authorize signer X (and the authorization is completed) for some role, and I then authorize signer Y for the same role and never complete that authorization.  The second authorization would overwrite the first, and leave that role without any authorized signer, right?  If that is so, this could be used as a hacky way to revoke an authorization without setting a new one, which raises the question whether we envision a separate function to revoke authorizations.

Reply:
Ohh bringing up revocation is a great one! I think revocation would be for sure great to support, and should be done explicitly IMO.

You scenario of an overwritten, but not completed signer is also a great point. IMO, we should allow for signers to be "updated atomically", even if completion of the authorization does not. What do you think?

Reply:
I'm not sure what you're suggesting.  The update being atomic is already the case if you use `authorizeSignerWithSignature()`, while if you use `authorizeSigner()` I'm not sure what it is that should be atomic or in what sense.

The only thing I could think of is maybe you mean for the old signer to be valid until the new signer authorization has been completed.  But I would hesitate to add any implementation or storage complexity for something like that, given that `authorizeSignerWithSignature()` is also an option for those who want 'zero-downtime' in signer authorization.

Reply:
Another thought here is expanding to allow multiple signers per `role`.

To me that seems definitely what we'd want for general roles, but perhaps `Attestation` , `Vote` and `Validator` roles should be special cased to match the current implementation (ie. just one signer).